Iceberg View Functionality Issues in SparkSQL (PySpark) and Trino

Two distinct issues have been identified concerning Iceberg view interactions across different query engines:

1.  **SparkSQL (PySpark):**
    * While standard Iceberg table interactions function as expected, the creation of Iceberg views is consistently failing.
2.  **Trino:**
    * Iceberg view creation and basic querying are successful.
    * However, the `CREATE OR REPLACE VIEW` statement fails to update existing views. A manual `DROP VIEW` followed by `CREATE VIEW` is required to achieve the desired update.

### Steps to Reproduce

**Trino:**

1.  Establish a Trino connection to our Lakekeeper server
2.  Create an Iceberg view.
3.  Attempt to replace the created view using `CREATE OR REPLACE VIEW`.
4.  Observe the failure of the replace operation.

**SparkSQL (PySpark):**

1.  Establish a SparkSession configured for Iceberg connect to our Lakekeeper server
2.  Attempt to create an Iceberg view using SparkSQL syntax (e.g., `spark.sql("CREATE VIEW ... AS SELECT ...")`). (Already drop view before do this command)
3.  Observe the failure while still can query table.

### Expected Behavior

Iceberg views should function consistently across different query engines, including:

* Successful creation from SparkSQL (PySpark).
* Proper execution of the `CREATE OR REPLACE VIEW` statement in Trino, enabling seamless view updates without manual dropping.

**Screenshots**
- Can't create view from Spark 
![Image](https://github.com/user-attachments/assets/d24240aa-3fc9-400d-a553-88ed510eca65)
- Can't replace existed view
![Image](https://github.com/user-attachments/assets/ea831e9a-ed9e-4b80-97c7-923176f78844)
![Image](https://github.com/user-attachments/assets/9f0eb83d-1991-41f3-afa0-d6f3b2dcc288)

### Environment
Spark: 3.5
Iceberg: 1.7.1
Authentication: keycloak
Trino: 455
Storage: S3

> You can set `"iceberg.unique-table-location" = 'true',` in your trino config, this will make table & view locations unique and prevents this issue.

By default, this property is enabled. I've already manually added it, but it still doesn't work. 

In fact, I'm getting a different error message, and I haven't found any solution for it:
`Caused by: org.apache.iceberg.exceptions.BadRequestException: Malformed request: Setting location for views is not supported`
![Image](https://github.com/user-attachments/assets/f33390ce-5bc4-4f19-ae89-c580454d066b)

Do you know what might be the issue? My configurations:
```yaml
  lakekeeper: |
    connector.name=iceberg
    iceberg.catalog.type=rest
    iceberg.rest-catalog.uri=https://lakekeeper.xyz/catalog
    iceberg.rest-catalog.warehouse=test-sts-bucket
    iceberg.rest-catalog.security=OAUTH2
    iceberg.rest-catalog.oauth2.credential=${ENV:LAKEKEEPER_CLIENT_ID}:${ENV:LAKEKEEPER_CLIENT_SECRET}
    iceberg.rest-catalog.vended-credentials-enabled=true
    iceberg.rest-catalog.oauth2.scope=lakekeeper
    iceberg.rest-catalog.oauth2.server-uri=https://keycloak.xyz/realms/iceberg/protocol/openid-connect/token
    iceberg.unique-table-location=true
    s3.region=us-east-2
    s3.path-style-access=true
    s3.endpoint=https://cephrgw.xyz
    fs.native-s3.enabled=true
```

> > You can set `"iceberg.unique-table-location" = 'true',` in your trino config, this will make table & view locations unique and prevents this issue.
> 
> By default, this property is enabled. I've already manually added it, but it still doesn't work.
> 
> In fact, I'm getting a different error message(I didn't specify location for table and view), and I haven't found any solution for it: `Caused by: org.apache.iceberg.exceptions.BadRequestException: Malformed request: Setting location for views is not supported`

This error is only produced when a view update happens which alters an existing view by changing its location:

```rust
            ViewUpdate::SetLocation { location } => {
                if previous_location != location {
                    return Err(ErrorModel::builder()
                        .code(StatusCode::BAD_REQUEST.into())
                        .message("Setting location for views is not supported".to_string())
                        .r#type("SetLocationNotSupported".to_string())
                        .build()
                        .into());
                }
                m
            }
```

Is it possible that you changed the unique table location setting which would alter the location trino forwards to lakekeeper when recreating the view?


> Hi @twuebi, thanks for your response above, Do you have any update on this as well?
@baotran306 could you create a second issue for this?

> I'm getting a bit lost here, could you:
> 
> 1. Restart Trino
> 2. Restart Jupyter or your Python environment
> 3. Share a minimal example script which runs into this issue similar to what I provided [here](https://github.com/lakekeeper/lakekeeper/issues/927#issuecomment-2743101214)
> 
> This will help me understand better what's happening and provide a better solution.

Sure, I've restarted and followed all these steps using Trino in DBeaver, and I'm still getting the same error, but only with the final `CREATE OR REPLACE` command.

`Caused by: org.apache.iceberg.exceptions.BadRequestException: Malformed request: Setting location for views is not supported`

```sql
-- Create table without location, insert data, and query
create table lakekeeper.risk_mon_out.test_table(
	my_ints INT, my_floats DOUBLE, strings VARCHAR 
) WITH (format='PARQUET');

INSERT INTO lakekeeper.risk_mon_out.test_table VALUES (1, 1.0, 'a'), (2, 2.0, 'b');

select * from lakekeeper.risk_mon_out.test_table;

-- Initial new view and query
create or replace view lakekeeper.risk_mon_out.test_new_view_trino as
	select strings, current_timestamp(6) cur_time, count(1) cnt
	from lakekeeper.risk_mon_out.test_table
	group by strings;

select * from lakekeeper.risk_mon_out.test_new_view_trino;

-- Try to replace, all the commands above were successful except this one.
create or replace view lakekeeper.risk_mon_out.test_new_view_trino as
	select strings, sum(my_floats) cnt
	from lakekeeper.risk_mon_out.my_table
	group by strings
	order by 2 desc;
```

![Image](https://github.com/user-attachments/assets/054ed4c5-6192-46a1-8e74-6113df3cebec)

--- a/authz/opa-bridge/policies/trino/allow_table.rego
+++ b/authz/opa-bridge/policies/trino/allow_table.rego
@@ -85,6 +85,15 @@ allow_table_metadata if {
     trino.require_table_access(catalog, schema, table, "get_metadata")
 }
 
+allow_table_metadata if {
+    input.action.operation in ["FilterTables", "ShowColumns", "FilterColumns"]
+    catalog := input.action.resource.table.catalogName
+    schema := input.action.resource.table.schemaName
+    table := input.action.resource.table.tableName
+    trino.require_view_access(catalog, schema, table, "get_metadata")
+}
+
+
 allow_table_read if {
     input.action.operation in ["SelectFromColumns"]
     catalog := input.action.resource.table.catalogName

--- a/authz/opa-bridge/policies/trino/allow_table.rego
+++ b/authz/opa-bridge/policies/trino/allow_table.rego
@@ -101,3 +101,12 @@ allow_table_read if {
     table := input.action.resource.table.tableName
     trino.require_table_access(catalog, schema, table, "read_data")
 }
+
+
+allow_table_read if {
+    input.action.operation in ["SelectFromColumns"]
+    catalog := input.action.resource.table.catalogName
+    schema := input.action.resource.table.schemaName
+    table := input.action.resource.table.tableName
+    trino.require_view_access(catalog, schema, table, "get_metadata")
+}

--- a/crates/iceberg-catalog/src/catalog/views/commit.rs
+++ b/crates/iceberg-catalog/src/catalog/views/commit.rs
@@ -1,5 +1,8 @@
 use iceberg::spec::{ViewFormatVersion, ViewMetadata, ViewMetadataBuilder};
-use iceberg_ext::catalog::{rest::ViewUpdate, ViewRequirement};
+use iceberg_ext::{
+    catalog::{rest::ViewUpdate, ViewRequirement},
+    configs::Location,
+};
 use uuid::Uuid;
 
 use crate::{
@@ -9,7 +12,7 @@ use crate::{
     },
     catalog::{
         compression_codec::CompressionCodec,
-        io::write_metadata_file,
+        io::{remove_all, write_metadata_file},
         require_warehouse_id,
         tables::{
             determine_table_ident, extract_count_from_metadata_location, maybe_body_to_json,
@@ -95,7 +98,7 @@ pub(crate) async fn commit_view<C: Catalog, A: Authorizer + Clone, S: SecretStor
         metadata_location: before_update_metadata_location,
         metadata: before_update_metadata,
     } = C::load_view(view_id, false, t.transaction()).await?;
-    let view_location = parse_view_location(before_update_metadata.location())?;
+    let before_update_view_location = parse_view_location(before_update_metadata.location())?;
     let before_update_metadata_location = parse_view_location(&before_update_metadata_location)?;
 
     state
@@ -108,8 +111,13 @@ pub(crate) async fn commit_view<C: Catalog, A: Authorizer + Clone, S: SecretStor
     // serialize body before moving it
     let body = maybe_body_to_json(&request);
 
-    let requested_update_metadata = build_new_metadata(request, before_update_metadata)?;
+    let (requested_update_metadata, delete_old_location) = build_new_metadata(
+        request,
+        before_update_metadata,
+        &before_update_view_location,
+    )?;
 
+    let view_location = parse_view_location(requested_update_metadata.location())?;
     let metadata_location = storage_profile.default_metadata_location(
         &view_location,
         &CompressionCodec::try_from_properties(requested_update_metadata.properties())?,
@@ -169,6 +177,17 @@ pub(crate) async fn commit_view<C: Catalog, A: Authorizer + Clone, S: SecretStor
         .await?;
     t.commit().await?;
 
+    if let Some(DeleteLocation(before_update_view_location)) = delete_old_location {
+        // we discard the error since the operation went through and leaving a stale metadata file
+        // seems acceptable, returning errors here will simply prompt a retry which means another
+        // commit will be attempted that would still not clean this left-behind file up.
+        tracing::debug!("Deleting old view location at: '{before_update_view_location}'");
+        let _ = remove_all(&file_io, before_update_view_location)
+            .await
+            .inspect(|()| tracing::trace!("Deleted old view location"))
+            .inspect_err(|e| tracing::error!("Failed to queue tabular purge: {:?}", e));
+    }
+
     let _ = state
         .v1_state
         .publisher
@@ -223,13 +242,17 @@ fn check_asserts(
     Ok(())
 }
 
+struct DeleteLocation<'c>(&'c Location);
+
 fn build_new_metadata(
     request: CommitViewRequest,
     before_update_metadata: ViewMetadata,
-) -> Result<ViewMetadata> {
+    before_location: &Location,
+) -> Result<(ViewMetadata, Option<DeleteLocation<'_>>)> {
     let previous_location = before_update_metadata.location().to_string();
-    let mut m = ViewMetadataBuilder::new_from_metadata(before_update_metadata);
 
+    let mut m = ViewMetadataBuilder::new_from_metadata(before_update_metadata);
+    let mut delete_old_location = None;
     for upd in request.updates {
         m = match upd {
             ViewUpdate::AssignUuid { .. } => {
@@ -241,15 +264,10 @@ fn build_new_metadata(
                 .into());
             }
             ViewUpdate::SetLocation { location } => {
-                if previous_location != location {
-                    return Err(ErrorModel::bad_request(
-                        "Setting location for views is not supported",
-                        "SetLocationNotSupported",
-                        None,
-                    )
-                    .into());
+                if location != previous_location {
+                    delete_old_location = Some(DeleteLocation(before_location));
                 }
-                m
+                m.set_location(location)
             }
 
             ViewUpdate::UpgradeFormatVersion { format_version } => match format_version {
@@ -295,7 +313,7 @@ fn build_new_metadata(
             Some(Box::new(e)),
         )
     })?;
-    Ok(requested_update_metadata.metadata)
+    Ok((requested_update_metadata.metadata, delete_old_location))
 }
 
 #[cfg(test)]

--- a/crates/iceberg-catalog/src/catalog/views/commit.rs
+++ b/crates/iceberg-catalog/src/catalog/views/commit.rs
@@ -125,6 +125,10 @@ pub(crate) async fn commit_view<C: Catalog, A: Authorizer + Clone, S: SecretStor
         extract_count_from_metadata_location(&before_update_metadata_location).map_or(0, |v| v + 1),
     );
 
+    if delete_old_location.is_some() {
+        storage_profile.require_allowed_location(&view_location)?;
+    }
+
     C::update_view_metadata(
         namespace_id,
         view_id,
@@ -185,7 +189,7 @@ pub(crate) async fn commit_view<C: Catalog, A: Authorizer + Clone, S: SecretStor
         let _ = remove_all(&file_io, before_update_view_location)
             .await
             .inspect(|()| tracing::trace!("Deleted old view location"))
-            .inspect_err(|e| tracing::error!("Failed to queue tabular purge: {:?}", e));
+            .inspect_err(|e| tracing::error!("Failed to delete old view location: {e:?}"));
     }
 
     let _ = state

--- a/tests/python/tests/conftest.py
+++ b/tests/python/tests/conftest.py
@@ -113,8 +113,8 @@ def storage_config(request) -> dict:
             pytest.skip("LAKEKEEPER_TEST__S3_BUCKET is not set")
 
         if (
-            settings.s3_path_style_access is not None
-            and settings.s3_path_style_access != ""
+                settings.s3_path_style_access is not None
+                and settings.s3_path_style_access != ""
         ):
             path_style_access = string_to_bool(settings.s3_path_style_access)
         else:
@@ -150,8 +150,8 @@ def storage_config(request) -> dict:
             pytest.skip("LAKEKEEPER_TEST__AWS_S3_BUCKET is not set")
 
         if (
-            settings.s3_path_style_access is not None
-            and settings.s3_path_style_access != ""
+                settings.s3_path_style_access is not None
+                and settings.s3_path_style_access != ""
         ):
             path_style_access = string_to_bool(settings.s3_path_style_access)
         else:
@@ -183,13 +183,13 @@ def storage_config(request) -> dict:
         }
     elif request.param["type"] == "azure":
         if (
-            settings.azure_storage_account_name is None
-            or settings.azure_storage_account_name == ""
+                settings.azure_storage_account_name is None
+                or settings.azure_storage_account_name == ""
         ):
             pytest.skip("LAKEKEEPER_TEST__AZURE_STORAGE_ACCOUNT_NAME is not set")
         if (
-            settings.azure_storage_filesystem is None
-            or settings.azure_storage_filesystem == ""
+                settings.azure_storage_filesystem is None
+                or settings.azure_storage_filesystem == ""
         ):
             pytest.skip("LAKEKEEPER_TEST__AZURE_STORAGE_FILESYSTEM is not set")
 
@@ -283,7 +283,7 @@ class Server:
         return uuid.UUID(project_id)
 
     def create_warehouse(
-        self, name: str, project_id: uuid.UUID, storage_config: dict
+            self, name: str, project_id: uuid.UUID, storage_config: dict
     ) -> uuid.UUID:
         """Create a warehouse in this server"""
 
@@ -512,8 +512,8 @@ def spark(warehouse: Warehouse, storage_config):
         f"spark.sql.catalog.{catalog_name}.oauth2-server-uri": f"{settings.token_endpoint}",
     }
     if (
-        storage_config["storage-profile"]["type"] == "s3"
-        and storage_config["storage-profile"]["sts-enabled"]
+            storage_config["storage-profile"]["type"] == "s3"
+            and storage_config["storage-profile"]["sts-enabled"]
     ):
         configuration[
             f"spark.sql.catalog.{catalog_name}.header.X-Iceberg-Access-Delegation"
@@ -589,7 +589,8 @@ def trino(warehouse: Warehouse, storage_config, trino_token):
             "iceberg.rest-catalog.oauth2.scope" = '{settings.openid_scope}',
             "iceberg.rest-catalog.oauth2.server-uri" = '{settings.token_endpoint}',
             "iceberg.rest-catalog.nested-namespace-enabled" = 'true',
-            "iceberg.rest-catalog.vended-credentials-enabled" = 'true'
+            "iceberg.rest-catalog.vended-credentials-enabled" = 'true',
+            "iceberg.unique-table-location" = 'true'
             {extra_config}
         )
     """

--- a/tests/python/tests/test_trino.py
+++ b/tests/python/tests/test_trino.py
@@ -142,7 +142,7 @@ def tets_create_view(trino, warehouse: conftest.Warehouse):
     assert r == [("a",), ("b",)]
 
 
-def tets_replace_view(trino, warehouse: conftest.Warehouse):
+def test_replace_view(trino, warehouse: conftest.Warehouse):
     ns = "tets_replace_view"
     cur = trino.cursor()
     cur.execute(f"CREATE SCHEMA {ns}")

--- a/tests/python/tests/test_trino.py
+++ b/tests/python/tests/test_trino.py
@@ -7,8 +7,8 @@ def test_create_namespace(trino, warehouse: conftest.Warehouse):
     cur = trino.cursor()
     cur.execute("CREATE SCHEMA test_create_namespace_trino")
     assert (
-        "test_create_namespace_trino",
-    ) in warehouse.pyiceberg_catalog.list_namespaces()
+               "test_create_namespace_trino",
+           ) in warehouse.pyiceberg_catalog.list_namespaces()
     schemas = cur.execute("SHOW SCHEMAS").fetchall()
     assert ["test_create_namespace_trino"] in schemas
 
@@ -27,20 +27,20 @@ def test_namespace_create_if_not_exists(trino, warehouse: conftest.Warehouse):
     cur.execute("CREATE SCHEMA IF NOT EXISTS test_namespace_create_if_not_exists_trino")
     cur.execute("CREATE SCHEMA IF NOT EXISTS test_namespace_create_if_not_exists_trino")
     assert (
-        "test_namespace_create_if_not_exists_trino",
-    ) in warehouse.pyiceberg_catalog.list_namespaces()
+               "test_namespace_create_if_not_exists_trino",
+           ) in warehouse.pyiceberg_catalog.list_namespaces()
 
 
 def test_drop_namespace(trino, warehouse: conftest.Warehouse):
     cur = trino.cursor()
     cur.execute("CREATE SCHEMA test_drop_namespace_trino")
     assert (
-        "test_drop_namespace_trino",
-    ) in warehouse.pyiceberg_catalog.list_namespaces()
+               "test_drop_namespace_trino",
+           ) in warehouse.pyiceberg_catalog.list_namespaces()
     cur.execute("DROP SCHEMA test_drop_namespace_trino")
     assert (
-        "test_drop_namespace_trino",
-    ) not in warehouse.pyiceberg_catalog.list_namespaces()
+               "test_drop_namespace_trino",
+           ) not in warehouse.pyiceberg_catalog.list_namespaces()
 
 
 def test_create_table(trino, warehouse: conftest.Warehouse):
@@ -89,9 +89,9 @@ def test_nested_schema(trino, warehouse: conftest.Warehouse):
     cur.execute("CREATE SCHEMA test_nested_schema_trino")
     cur.execute('CREATE SCHEMA "test_nested_schema_trino.nested"')
     assert (
-        "test_nested_schema_trino",
-        "nested",
-    ) in warehouse.pyiceberg_catalog.list_namespaces(
+               "test_nested_schema_trino",
+               "nested",
+           ) in warehouse.pyiceberg_catalog.list_namespaces(
         "test_nested_schema_trino",
     )
 
@@ -117,9 +117,9 @@ def test_rename_table(trino, warehouse: conftest.Warehouse):
         "ALTER TABLE test_rename_table_trino.my_table RENAME TO test_rename_table_trino.my_table_renamed"
     )
     assert (
-        "test_rename_table_trino",
-        "my_table_renamed",
-    ) in warehouse.pyiceberg_catalog.list_tables("test_rename_table_trino")
+               "test_rename_table_trino",
+               "my_table_renamed",
+           ) in warehouse.pyiceberg_catalog.list_tables("test_rename_table_trino")
 
 
 def tets_create_view(trino, warehouse: conftest.Warehouse):
@@ -140,3 +140,28 @@ def tets_create_view(trino, warehouse: conftest.Warehouse):
     )
     r = cur.execute("SELECT * FROM test_create_view_trino.my_view").fetchall()
     assert r == [("a",), ("b",)]
+
+
+def tets_replace_view(trino, warehouse: conftest.Warehouse):
+    ns = "tets_replace_view"
+    cur = trino.cursor()
+    cur.execute(f"CREATE SCHEMA {ns}")
+    cur.execute(
+        f"CREATE TABLE {ns}.my_table (my_ints INT, my_floats DOUBLE, strings VARCHAR) WITH (format='PARQUET')"
+    )
+    cur.execute(
+        f"CREATE OR REPLACE VIEW {ns}.my_view AS SELECT strings FROM {ns}.my_table"
+    )
+    assert ("my_view",) in warehouse.pyiceberg_catalog.list_tables(
+        ns
+    )
+    # Insert data and query view
+    cur.execute(
+        f"INSERT INTO {ns}.my_table VALUES (1, 1.0, 'a'), (2, 2.0, 'b')"
+    )
+    r = cur.execute(f"SELECT * FROM {ns}.my_view").fetchall()
+    assert r == [("a",), ("b",)]
+
+    cur.execute(
+        f"CREATE OR REPLACE VIEW {ns}.my_view AS SELECT strings FROM {ns}.my_table"
+    )

--- a/tests/python/tests/test_trino.py
+++ b/tests/python/tests/test_trino.py
@@ -1,6 +1,4 @@
 import conftest
-import pandas as pd
-import pytest
 
 
 def test_create_namespace(trino, warehouse: conftest.Warehouse):
@@ -142,8 +140,8 @@ def tets_create_view(trino, warehouse: conftest.Warehouse):
     assert r == [("a",), ("b",)]
 
 
-def test_replace_view(trino, warehouse: conftest.Warehouse):
-    ns = "tets_replace_view"
+def test_replace_view(trino):
+    ns = "test_replace_view"
     cur = trino.cursor()
     cur.execute(f"CREATE SCHEMA {ns}")
     cur.execute(
@@ -152,9 +150,7 @@ def test_replace_view(trino, warehouse: conftest.Warehouse):
     cur.execute(
         f"CREATE OR REPLACE VIEW {ns}.my_view AS SELECT strings FROM {ns}.my_table"
     )
-    assert ("my_view",) in warehouse.pyiceberg_catalog.list_tables(
-        ns
-    )
+    assert ["my_view"] in cur.execute(f"SHOW TABLES IN {ns}").fetchall()
     # Insert data and query view
     cur.execute(
         f"INSERT INTO {ns}.my_table VALUES (1, 1.0, 'a'), (2, 2.0, 'b')"

--- a/tests/python/tests/test_trino.py
+++ b/tests/python/tests/test_trino.py
@@ -120,7 +120,7 @@ def test_rename_table(trino, warehouse: conftest.Warehouse):
            ) in warehouse.pyiceberg_catalog.list_tables("test_rename_table_trino")
 
 
-def tets_create_view(trino, warehouse: conftest.Warehouse):
+def test_create_view(trino, warehouse: conftest.Warehouse):
     cur = trino.cursor()
     cur.execute("CREATE SCHEMA test_create_view_trino")
     cur.execute(
@@ -129,15 +129,14 @@ def tets_create_view(trino, warehouse: conftest.Warehouse):
     cur.execute(
         "CREATE OR REPLACE VIEW test_create_view_trino.my_view AS SELECT strings FROM test_create_view_trino.my_table"
     )
-    assert ("my_view",) in warehouse.pyiceberg_catalog.list_tables(
-        "test_create_view_trino"
-    )
+    assert ["my_view"] in cur.execute(f"SHOW TABLES IN test_create_view_trino").fetchall()
+
     # Insert data and query view
     cur.execute(
         "INSERT INTO test_create_view_trino.my_table VALUES (1, 1.0, 'a'), (2, 2.0, 'b')"
     )
     r = cur.execute("SELECT * FROM test_create_view_trino.my_view").fetchall()
-    assert r == [("a",), ("b",)]
+    assert r == [["a"], ["b"]]
 
 
 def test_replace_view(trino):
@@ -156,7 +155,7 @@ def test_replace_view(trino):
         f"INSERT INTO {ns}.my_table VALUES (1, 1.0, 'a'), (2, 2.0, 'b')"
     )
     r = cur.execute(f"SELECT * FROM {ns}.my_view").fetchall()
-    assert r == [("a",), ("b",)]
+    assert r == [["a"], ["b"]]
 
     cur.execute(
         f"CREATE OR REPLACE VIEW {ns}.my_view AS SELECT strings FROM {ns}.my_table"

